torch/testing/_internal/common_utils.py

```diff
@@ -2165,13 +2165,22 @@ def _run_with_retry(self, result=None, num_runs_left=0, report_only=True, num_re
                 result.addExpectedFailure(self, err)
             self._run_with_retry(result=result, num_runs_left=num_retries_left, report_only=report_only,
                                  num_red=num_red + 1, num_green=num_green)
-        elif (RERUN_DISABLED_TESTS or report_only) and num_retries_left < MAX_NUM_RETRIES:
-            # Always re-run up to MAX_NUM_RETRIES when running under report only or rerun disabled tests modes
+        elif RERUN_DISABLED_TESTS and num_retries_left <= MAX_NUM_RETRIES and not result.skipped:
+            # Always re-run up to MAX_NUM_RETRIES when running under rerun disabled tests modes if the test successes.
+            # The parameter num_retries_left can be equal to MAX_NUM_RETRIES here because num_runs_left is initially
+            # set to MAX_NUM_RETRIES + 1, i.e. the first run successes
+            #
+            # Also if the result is skipped, this is due to check_if_enable skipping non-disabled tests, thus we
+            # want to ignore them, not retrying and skipping multiple times
             print(f"    {self._testMethodName} succeeded - num_retries_left: {num_retries_left}")
-            if RERUN_DISABLED_TESTS:
-                result.addSuccess(self)
-            else:
-                result.addUnexpectedSuccess(self)
+            result.addSuccess(self)
+            self._run_with_retry(result=result, num_runs_left=num_retries_left, report_only=report_only,
+                                 num_red=num_red, num_green=num_green + 1)
+        elif report_only and num_retries_left < MAX_NUM_RETRIES:
```

Comment:
This is the original logic in this branch before https://github.com/pytorch/pytorch/pull/88646